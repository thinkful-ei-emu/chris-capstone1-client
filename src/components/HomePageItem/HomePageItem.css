.HomePageItem {
    background-color: #CC8B56;
    display: flex;
    flex-direction: column;
    border: 0;
    text-decoration: none;
    box-shadow: 0px 11px 21px -8px rgba(0, 0, 0, 0.5);
    margin-bottom: 40px;
    margin-right: auto;
    margin-left: auto;
    border-radius: 10px;
    padding: 5px;
  }
  
  .HomePageItem:hover {
    cursor: pointer;
  }
  
  .HomePageItem_image {
    height: 256px;
    width: 160px;
    border-radius: 10px 10px 0 0;
    background-size: cover;
    background-position: 50% 50%;
  }
  
  .HomePageItem_details {
    padding: 10px;
    display: flex;
    flex: 1;
    flex-direction: column;
    justify-content: space-between;
    width: 250px;
  }
  
  .HomePageItem_details h2 {
    margin: 0;
  }
  
  .HomePageItem_details span {
    display: block;
    text-align: center;
    opacity: 0.5;
  }
  
  .addRating {
    width: 50px;
    display: inline;
  }