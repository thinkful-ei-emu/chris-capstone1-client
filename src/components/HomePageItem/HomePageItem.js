import React from 'react';
import { Button, Input } from '../Utils/Utils';
import BookApiService from '../../services/book-api-service';
import './HomePageItem.css';

export default class HomePageItem extends React.Component {
    state = {
        error: null,
        rating: undefined,
        ratedAlready: false,
        ratingId: undefined
    }

    componentDidMount(){
        //if this has been rated by the person already, allow them to delete their rating
        BookApiService.getRating()
        .then(res => res.find(rating => rating.book_id === this.props.book.id))
        .then(res => {
            if(!!res)
                this.setState({ ratedAlready: !!res, ratingId: res.id })
        })
    }


    updateBookRating = rating => this.setState({ rating })

    renderDeleteRating(){
        const { removeRating } = this.props;
        const { ratingId } = this.state;
        return<>
            <Button onClick={async () => {
                await removeRating(ratingId)
                this.setState({ ratedAlready: false, ratingId: undefined })
                }}>
                Delete Rating
            </Button>
            </>
    }

    renderAddRating(){
        const {book, addRating } = this.props;
        return<div>
            <p>Read it before? What's your grade?</p>
        <Input className='addRating' type='number' value={this.state.rating} onChange={e => this.updateBookRating(e.target.value)} />
        <span><strong>% </strong></span>
            <Button onClick={async () => {
                await addRating(this.state.rating, book.id)
                await BookApiService.getRating()
                    .then(res => res.find(rat => rat.book_id === this.props.book.id))
                    .then(res => {
                        if(!!res)
                            this.setState({ ratedAlready: !!res, ratingId: res.id })
                    })
                    .catch(error => console.error(error))
                this.setState({ ratedAlready: true })
            }}>
                Grade!
            </Button>
            </div>
    }

    render () {
        const {book, addToYourShelf} = this.props;
        const newBook = {
            title: book.title,
            author: book.author,
            book_source: book.book_source,
            image: book.image,
            list: 'wishlist',
            genre: book.genre
        }

        return (
            <div className='HomePageItem'>
            {book.image && <img className='HomePageItem_image' src={book.image} alt='the cover of the book' />}

            <div className='HomePageItem_details'>
                <div className='HomePageItem_text'>
                    <h2 className='HomePageItem_title'>{book.title}</h2>
                    <p className='HomePageItem_author'>{book.author}</p>
                </div>
            </div>
            <div>
            <Button onClick={() => addToYourShelf(newBook)}>
                Add to Your Shelf
            </Button>
            {' '}
            {this.state.ratedAlready 
              ? this.renderDeleteRating()
              : this.renderAddRating()}
              </div>
        </div>
        )
    }
}