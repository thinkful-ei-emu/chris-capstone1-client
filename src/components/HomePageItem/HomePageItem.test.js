import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import HomePageItem from './HomePageItem';
import ReactDom from 'react-dom';

describe('HomePageItem component', () => {
  let book = {
    "id": 1,
    "title": "Disappearing Earth",
    "author": "Julia Phillips",
    "list": "wishlist",
    "date_created": "2019-08-09T16:06:49.108Z",
    "book_source": "Print",
    "image": "https://media.vanityfair.com/photos/5d1662c85f741a0008ed05e1/master/w_800%2Cc_limit/best-books-of-2019-Disappearing-Earth.jpg",
    "book_report": "",
    "rating": 13,
    "genre": "Fairytale",
    "recommended": false,
    "user": {
        "id": 4,
        "user_name": "s.smith",
        "full_name": "Sam Smith",
        "email": "Sam@gmail.com",
        "date_created": "2019-08-09T16:06:49.108Z"
    },
    "poster": {
        "poster_report": null
    },
    "number_of_ratings": 5,
    "average_rating": 69.8
}
    it('renders without crashing', () => {
        const div = document.createElement('div')
        ReactDom.render(
            <HomePageItem key={book.id} book={book} />,
          div
        )
        ReactDom.unmountComponentAtNode(div)
      })
    it('renderes HomePageItem by default', () => {
        expect(toJson(shallow(<HomePageItem key={book.id} book={book} />))).toMatchSnapshot() 
    })
})