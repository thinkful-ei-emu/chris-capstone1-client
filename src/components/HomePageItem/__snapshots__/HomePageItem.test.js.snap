// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePageItem component renderes HomePageItem by default 1`] = `
<div
  className="HomePageItem"
>
  <img
    alt="the cover of the book"
    className="HomePageItem_image"
    src="https://media.vanityfair.com/photos/5d1662c85f741a0008ed05e1/master/w_800%2Cc_limit/best-books-of-2019-Disappearing-Earth.jpg"
  />
  <div
    className="HomePageItem_details"
  >
    <div
      className="HomePageItem_text"
    >
      <h3
        className="HomePageItem_title"
      >
        Disappearing Earth
      </h3>
      <p
        className="HomePageItem_author"
      >
        by: 
        Julia Phillips
      </p>
      <p>
        Average User Rating: 
        <span>
          69.8
          %
        </span>
      </p>
    </div>
  </div>
  <div>
    <Button
      onClick={[Function]}
    >
      Add to Wishlist
    </Button>
     
    <div>
      <p>
        Read it? What's your grade? (1-100)
      </p>
      <Input
        className="addRating"
        onChange={[Function]}
        type="number"
      />
      <span>
        <strong>
          % 
        </strong>
      </span>
      <Button
        onClick={[Function]}
      >
        Grade!
      </Button>
    </div>
  </div>
</div>
`;
